### PR TITLE
ISPN-11226 Set the cache manager in the task context for distributed …

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-11226

This has the side-effect of being able to run distributed tasks in the server. To simplify things, only string parameters are supported for know. This limitation will be removed once we have registration of protobuf serialization contexts in the server.

### CODE REVIEW

**Comment by danberindei** [comment]
The changes look good to me, but I think something is missing from the test:
```
org.infinispan.client.hotrod.exceptions.HotRodClientException:Request for messageId=1450 returned server error (status=0x85): org.infinispan.commons.marshall.MarshallingException: No marshaller registered for Java type org.infinispan.server.tasks.DistributedServerTask
```

**Comment by tristantarrant** [comment]
Yep, that is why it is preview :)
I also found that the current implementation of server tasks is not thread-safe...

File: commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java [hunk 1]
```diff
@@ -78,10 +78,12 @@
    int FILE_READ_LOCK_KEY = LUCENE_LOWER_BOUND + 4;
    int FILE_LIST_CACHE_VALUE = LUCENE_LOWER_BOUND + 5;
 
-   // Scripting 4800 -> 4999
+   // Tasks + Scripting 4800 -> 4999
    int SCRIPTING_LOWER_BOUND = LUCENE_LOWER_BOUND + 200;
    int EXECUTION_MODE = SCRIPTING_LOWER_BOUND;
    int SCRIPT_METADATA = SCRIPTING_LOWER_BOUND + 1;
+   int DISTRIBUTED_SERVER_TASK = SCRIPT_METADATA + 1;
+   int DISTRIBUTED_SERVER_TASK_PARAMETER = DISTRIBUTED_SERVER_TASK + 1;
```
Comment (ryanemerson) [inline commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: This is much simpler than +5 etc. I'll update the class to use this approach for all values in the future.

File: server/runtime/src/main/java/org/infinispan/server/Extensions.java [hunk 2]
```diff
@@ -84,6 +84,7 @@ private void loadServerTasks(ClassLoader classLoader) {
    }
 
    public TaskEngine getServerTaskEngine(EmbeddedCacheManager cm) {
+      // Create marshallers for all tasks
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/Extensions.java]: Is this an old comment? It doesn't seem to belong here anymore.

File: server/runtime/src/main/java/org/infinispan/server/tasks/DistributedServerTaskRunner.java [hunk 3]
```diff
@@ -36,8 +38,9 @@ public DistributedServerTaskRunner(ServerTaskEngine serverTaskEngine) {
             results.add(v);
          }
       };
-      CompletableFuture<Void> future = clusterExecutor.submitConsumer(new DistributedServerTask<>(
-            masterCacheNode.getName(), taskName, context.getParameters()), triConsumer);
+      List<TaskParameter> taskParams = context.getParameters().orElse(Collections.emptyMap()).entrySet().stream().map(e -> new TaskParameter(e.getKey(), e.getValue().toString())).collect(Collectors.toList());
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/tasks/DistributedServerTaskRunner.java]: Class variable `serverTaskEngine` is never used, so we can remove the constructor.

File: server/runtime/src/main/java/org/infinispan/server/tasks/PersistenceContextInitializer.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ */
+@AutoProtoSchemaBuilder(
+      includeClasses = {
+            TaskParameter.class,
+            DistributedServerTask.class
+      },
+      schemaFileName = "persistence.servertasks.proto",
+      schemaFilePath = "proto/generated",
+      schemaPackageName = "org.infinispan.persistence.servertasks")
+public interface PersistenceContextInitializer extends SerializationContextInitializer {
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/tasks/PersistenceContextInitializer.java]: Can be package-private

File: server/runtime/src/main/java/org/infinispan/server/tasks/TaskParameter.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+@ProtoTypeId(ProtoStreamTypeIds.DISTRIBUTED_SERVER_TASK_PARAMETER)
+public class TaskParameter {
+   @ProtoField(number = 1)
+   String key;
+
+   @ProtoField(number = 2)
+   String value;
+
+   @ProtoFactory
+   public TaskParameter(String key, String value) {
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/tasks/TaskParameter.java]: Class and constructor can be package-private as only used internally.

File: commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java (hunk 1 above)
Comment (wburns) [reply commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: You just can't comment them out arbitrarily anymore though :D

File: server/runtime/src/main/java/org/infinispan/server/tasks/TaskParameter.java (hunk 5 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/tasks/TaskParameter.java]: Fixed

File: server/runtime/src/main/java/org/infinispan/server/tasks/PersistenceContextInitializer.java (hunk 4 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/tasks/PersistenceContextInitializer.java]: Done

File: server/runtime/src/main/java/org/infinispan/server/tasks/DistributedServerTaskRunner.java (hunk 3 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/tasks/DistributedServerTaskRunner.java]: Removed

File: server/runtime/src/main/java/org/infinispan/server/Extensions.java (hunk 2 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/Extensions.java]: Removed

**Comment by ryanemerson** [comment]
@tristantarrant Checkstyle failures on CI.

**Comment by tristantarrant** [comment]
Fixed now, @ryanemerson

**Comment by ryanemerson** [comment]
Thanks @tristantarrant

**Pull request merged**